python_modules/dagster/dagster/components/definitions.py

```diff
@@ -109,3 +115,99 @@ def regular_asset(): ...
+    """
+
+    @abc.abstractmethod
+    def map_asset_specs(
+        self: T,
+        *,
+        func: Callable[[AssetSpec], AssetSpec],
+        selection: Optional[CoercibleToAssetSelection] = None,
+    ) -> T:
+        """Map a function over the included AssetSpecs or AssetsDefinitions in this Definitions object, replacing specs in the sequence
+        or specs in an AssetsDefinitions with the result of the function.
+
+        Args:
+            func (Callable[[AssetSpec], AssetSpec]): The function to apply to each AssetSpec.
+            selection (Optional[Union[str, Sequence[str], Sequence[AssetKey], Sequence[Union[AssetsDefinition, SourceAsset]], AssetSelection]]): An asset selection to narrow down the set of assets to apply the function to. If not provided, applies to all assets.
+
+        Returns:
+            DefinitionsHandle: A DefinitionsHandle object where the AssetSpecs have been replaced with the result of the function where the selection applies.
+
+        Examples:
+            .. code-block:: python
+
+                defs_handle: DefinitionsHandle = ...
+
+                # Applies to asset1 and asset2
+                mapped_def_handle = defs_handle.map_asset_specs(
+                    func=lambda s: s.merge_attributes(metadata={"new_key": "new_value"}),
+                )
+
+        """
+        ...
+
+
+PostProcessorFn: TypeAlias = Callable[[DefinitionsHandle], DefinitionsHandle]
+
+
+class ComponentsDefinitionsHandle(DefinitionsHandle):
+    """Interface which sits on top of a `Definitions` object and allows modification of the contained
+    defs, with the restriction that transformations must be map-like. This ensures that the transformations
+    are identical in the case that we may subset the input Definitions object.
+
+    In the future, this class will cache asset selections on the load context so that subsequent loads
+    (e.g. in runs, instigators) can reuse the same asset selection without resolving the entire asset graph.
+    """
+
+    def __init__(self, defs: Definitions, context: ComponentLoadContext):
+        self._defs = defs
+        self._context = context
+
+    def map_asset_specs(
+        self,
+        *,
+        func: Callable[[AssetSpec], AssetSpec],
+        selection: Optional[CoercibleToAssetSelection] = None,
+    ) -> "ComponentsDefinitionsHandle":
+        if selection:
+            if isinstance(selection, str):
+                resolved_selection = AssetSelection.from_string(selection, include_sources=True)
+            else:
+                resolved_selection = AssetSelection.from_coercible(selection)
+        else:
+            resolved_selection = None
+        return self._map_asset_specs_inner(func, resolved_selection)
+
+    def _map_asset_specs_inner(
+        self, func: Callable[[AssetSpec], AssetSpec], selection: Optional[AssetSelection]
+    ) -> "ComponentsDefinitionsHandle":
+        # TODO: resolve target keys from cache in subsequent subset loads
```

Comment:
The idea here is that the asset selection is only resolvable on a full asset graph. On initial load, we resolve this selection as expected, and stash the list of keys. On subsequent (possibly subset) loads we can just load `target_keys` from the cache.

![image](https://github.com/user-attachments/assets/1a5fb068-600a-46d0-b2db-1a30bcd3d0d4)